torch/utils/data/_utils/pin_memory.py

```diff
@@ -74,6 +74,6 @@ def pin_memory(data, device=None):
             # The sequence type may not support `__init__(iterable)` (e.g., `range`).
             return [pin_memory(sample, device) for sample in data]
     elif hasattr(data, "pin_memory"):
-        return data.pin_memory()
+        return data.pin_memory(device)
```

Comment:
I wonder if some hacks like checking method signature are needed, as probably for many years implementors assumed and implemented an argument-less variant

Reply:
Sure, I will try to add those checks

Reply:
@vadimkantorov I have added some basic checks, please advise

Reply:
I'm not maintainer :) We need to wait for some real reviewers. The above is just a consideration that seems reasonable. But I don't have a larger design picture in mind